### PR TITLE
Add support single sign-on logins through SAML 2.0

### PR DESCRIPTION
**Pull request by diego-araujo**
Prezados,

Envio contribuição que adiciona outro meio de autenticação (saml 2.0), os testes foram feitos em integração com um provedor de Identidades Shibboleth.
A solução foi necessária devido a requisitos da Universidade de São Paulo, que irá utilizar o Tsuru como solução para prover PaaS em nossa cloud.
Esta funcionalidade possui como dependência a lib  [XMLSec](https://www.aleksey.com/xmlsec/).  e, portanto, gostaria de saber como incluir uma dependência de terceiros?
Caso tenha algo fora do padrão, errado ou sugestões de melhorias, por gentileza me fiquem à vontade em relatar, assim poderemos alterar e contribuir na forma correta.

## Third-party dependencies

For enable SAMl authentication the library XMLSec is required.
You need install this in the API Server.
Ubuntu users can install by using `apt-get install xmlsec1`.
For others Linux Distributions you can download and compile the source code from [XMLSec](https://www.aleksey.com/xmlsec/download.html). 
Do not forget to install the required libraries first: libxml2, libxslt, iconv and zlib

## tsuru.conf reference

`auth:saml:request-expire-seconds.`
This  setting defines the amount of seconds that the auth request will be valid waiting for callback response from IDP. 
This setting is optional, and defaults to “180”.

`auth:saml:idp-attribute-user-identity`
This setting indicates the attribute name to be looked for in SAML response that MUST contain a unique user identifier
For internal purposes, this user identifier will be converted to a string email format, if necessary will be added `<user identifier>@<host from auth:saml:idp-ssourl>`

`auth:saml:sp-publiccert`
Tsuru X509 certificate file location

`auth:saml:sp-privatekey`
Tsuru private key file location

Here's a convenient way to generate a self-signed certificate :)

<pre>curl -sSL https://raw.githubusercontent.com/frntn/x509-san/master/gencert.sh | CRT_CN="mycert"  bash</pre>

`auth:saml:idp-ssourl`
The URL at the Identity Provider to which SAML authentication requests should be sent.

`auth:saml:idp-publiccert:`
Identity Provider public key certificate file location

`auth:saml:sp-entityid`
Indicate a unique identifier configured in your Identity Provider, typically the name of your domain (including a trailing slash).
    
`auth:saml:sp-display-name`
This filed specifies a localized name fit for display to users in your Identity Provider.
This setting is optional, and defaults to "Tsuru".

`auth:saml:sp-description`
This filed specifies a localized description fit for display to users in your Identity Provider.
This SHOULD be a description of the service being offered. 
This setting is optional, and defaults to "Tsuru Platform as a Service software".

`auth:saml:sp-sign-request`
Enables signing of requests sent to the IDP

`auth:saml:idp-sign-response`
This flag indicates whether the identity provider are signing the SAML response, api will check

`auth:saml:idp-deflate-encoding`
This flag indicates whether the Identity Provider response are compressed 

### Callback URL

SAML api callback URL expects to receive the messages via the HTTP POST binding from the Identity Provider
Callback URL: `<host>:<port>/auth/saml`
The SAML METADATA is available at the same URL `<host>:<port>/auth/saml`, but will only return metadata from HTTP GET.

Here is a example to configure saml auth:
```yaml
scheme: saml
  saml:
    request-expire-seconds: 180
    idp-attribute-user-identity: eduPersonPrincipalName
    sp-publiccert: <your-public-cert-file>
    sp-privatekey: <your-private-key-file>
    idp-ssourl: <http://idp.myprovider.com/Redirect/SSO>
    idp-publiccert: <idp-public-cert-file>
    sp-entityid: <your-entity-id>`
    sp-display-name: Tsuru PaaS Login
    sp-description: Tsuru Platform as a Service software
    sp-sign-request: true
    idp-sign-response: true
    idp-deflate-encoding: false

### CODE REVIEW

**Comment by cezarsa** [comment]
Hello Diego, thanks a lot for contributing!

Can you take a look at the failures reported by our [CI server](https://travis-ci.org/tsuru/tsuru/jobs/96029524)?

One of the problems is probably related to the missing `xmlsec1` library, you can fix this by adding a new entry to the `addons:apt:packages` section in the `tsuru.yaml` file.

There are also failures related to [code formatting](https://travis-ci.org/tsuru/tsuru/jobs/96029524#L273-L300) and you can fix them by running `goimports` on your code.

**Comment by diego-araujo** [comment]
Hi Cezar,

Thanks for your prompt response, now has been fixed.
This is my first go code, therefore any feedback is welcome!

File: api/saml.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2014 tsuru authors. All rights reserved.
```
Comment (andrewsmedina) [inline api/saml.go]: The copyright year should be 2015

File: api/saml.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+func samlMetadata(w http.ResponseWriter, r *http.Request) error {
+	if app.AuthScheme.Name() != "saml" {
+		return &errors.HTTP{
+			Code:    http.StatusBadRequest,
+			Message: "This URL is only supported with saml enabled",
+		}
+	}
+
```
Comment (andrewsmedina) [inline api/saml.go]: Can you remove the empty line inside the functions?

File: auth/saml/request.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+type Request struct {
+	ID       string    `json:"id"`
+	Creation time.Time `json:"creation"`
+	Expires  time.Time `json:"expires"`
+	Email    string    `json:"email"`
+	Authed   bool      `json:"authed"`
+}
+
+func loadExpireTime() error {
```
Comment (andrewsmedina) [inline auth/saml/request.go]: This function will never returns an error. I believe that a better approach is remove the global vars `requestExpire` and `defaultRequestExpiration` and change this function to returns the expire time instead to load it in the global vars.

File: auth/saml/request.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	_, err = conn.SAMLRequests().RemoveAll(bson.M{"id": req.ID})
+	return err
+
+}
+
+func removeOldTRequests() error {
```
Comment (andrewsmedina) [inline auth/saml/request.go]: The `OldT` is a typo?

File: db/storage.go [hunk 5]
```diff
@@ -158,6 +158,14 @@ func (s *Storage) Quota() *storage.Collection {
 	return c
 }
 
+// SAMLRequests returns the samlrequest from MongoDB.
```
Comment (andrewsmedina) [inline db/storage.go]: // SAMLRequests returns the saml_requests collections from MongoDB.
shoud be a better comment

File: auth/saml/request.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	"gopkg.in/mgo.v2/bson"
+)
+
+var (
+	requestExpire            time.Duration
+	defaultRequestExpiration = 3 * 60 * time.Second
+	ErrRequestNotFound       = &errors.ValidationError{Message: "request not found or expired"}
+)
+
+type Request struct {
```
Comment (andrewsmedina) [inline auth/saml/request.go]: the `request` can be `private`. it is used only by the `saml` package.

File: auth/saml/request.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	}
+	defer conn.Close()
+	err = conn.SAMLRequests().Find(bson.M{"id": id}).One(&request)
+	if err != nil {
+		return request, ErrRequestNotFound
+	}
+	return request, nil
+}
+
+func (req *Request) GetEmail() string {
```
Comment (andrewsmedina) [inline auth/saml/request.go]: the `GetEmail`, `SetAuth` and `SetEmail` getters/setters is not needed. You can access the `request` directly.

File: auth/saml/response.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/tsuru/config"
+	"github.com/tsuru/tsuru/errors"
+)
+
+var (
+	ErrRequestIdNotFound = &errors.ValidationError{Message: "Field InResponseTo not found in saml response data"}
+	ErrCheckSignature    = &errors.ValidationError{Message: "SAMLResponse signature validation"}
+)
+
+type Response struct {
```
Comment (andrewsmedina) [inline auth/saml/response.go]: the `Response` type can be `private`. it is used only by the `saml` package.

File: cmd/saml.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+		return data["token"].(string), nil
+	}
+
+	// finish when timeout
+	return "", saml.ErrRequestWaitingForCredentials
+}
+
+func (c *login) samlLogin(context *Context, client *Client) error {
+	//loadExpireTime()
```
Comment (andrewsmedina) [inline cmd/saml.go]: this line should be removed

File: auth/saml/saml_test.go [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func (s *S) TestSamlAuth(c *check.C) {
+
+	user := auth.User{Email: "x@x.com"}
+	token, _ := createToken(&user)
+	scheme := SAMLAuthScheme{}
+	strtoken, err := scheme.Auth("bearer " + token.GetValue())
+	c.Assert(err, check.IsNil)
+	//c.Assert(len(s.reqs), check.Equals, 1)
```
Comment (andrewsmedina) [inline auth/saml/saml_test.go]: this line should be removed

File: auth/saml/saml_test.go [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+	//c.Assert(s.reqs[0].URL.Path, check.Equals, "/user")
```
Comment (andrewsmedina) [inline auth/saml/saml_test.go]: this line should be removed

File: cmd/saml.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+	return successSamlCallBackMarkup
+}
+
+func SamlCallbackFailureMessage() string {
+	return errorSamlCallBackMarkup
+}
+
+func samlRequestId(schemeData map[string]string) string {
+	p := schemeData["request_id"]
+	return fmt.Sprintf("%s", p)
```
Comment (andrewsmedina) [inline cmd/saml.go]: the `fmt.Sprintf` is unecessary here. You can just return `schemeData["request_id"]`

File: cmd/saml.go [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+		t, err := t.Parse(formPostPage)
+		if err != nil {
+			page := fmt.Sprintf(errorSamlCallBackMarkup, err.Error())
+			w.Header().Add("Content-Type", "text/html")
+			w.Write([]byte(page))
+		} else {
+			t.Execute(w, schemeData)
+		}
+
+		return
```
Comment (andrewsmedina) [inline cmd/saml.go]: this `return` is unecessary

**Comment by diego-araujo** [comment]
Hi Andrews,
Thanks for the tips, all changes were implemented

File: api/saml.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+package api
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/tsuru/tsuru/app"
+	"github.com/tsuru/tsuru/auth"
+	saml "github.com/tsuru/tsuru/auth/saml"
```
Comment (fsouza) [inline api/saml.go]: Could you remove this import alias? The package is already named saml.

File: auth/saml/response.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package saml
+
+import (
+	"fmt"
+	"time"
+
+	saml "github.com/diego-araujo/go-saml"
```
Comment (fsouza) [inline auth/saml/response.go]: Can you remove this import alias as well?

File: auth/saml/saml.go [hunk 16]
```diff
@@ -0,0 +1,381 @@
...
+// license that can be found in the LICENSE file.
+
+package saml
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	saml "github.com/diego-araujo/go-saml"
```
Comment (fsouza) [inline auth/saml/saml.go]: ditto

File: auth/saml/saml.go [hunk 17]
```diff
@@ -0,0 +1,381 @@
...
+	}
+	return md, nil
+}
+
+func (s *SAMLAuthScheme) Login(params map[string]string) (auth.Token, error) {
+	_, err := s.loadConfig()
+	if err != nil {
+		return nil, err
+	}
+	//verify for callback requests, param 'callback' indicate callback
```
Comment (fsouza) [inline auth/saml/saml.go]: Can you remove this comment?

Also, I think we can make the whole condition smaller by combining if and assignment:

```golang
if _, ok := params["callback"]; ok {
    return nil, s.callback(params)
}
```

What do you think?

File: auth/saml/saml.go [hunk 18]
```diff
@@ -0,0 +1,381 @@
...
+	sp, err := s.createSP()
+	if err != nil {
+		return nil, err
+	}
+	// generate the AuthnRequest and then get a base64 encoded string of the XML
+	authnRequest := sp.GetAuthnRequest()
+	//b64XML, err := authnRequest.String(authnRequest)
+	b64XML, err := authnRequest.CompressedEncodedSignedString(sp.PrivateKeyPath)
+	//b64XML, err := authnRequest.EncodedSignedString(sp.PrivateKeyPath)
+	if err != nil {
```
Comment (fsouza) [inline auth/saml/saml.go]: Please remove these comments.

File: auth/saml/saml.go [hunk 19]
```diff
@@ -0,0 +1,381 @@
...
+	return &sp, nil
+}
+
+func (s *SAMLAuthScheme) Info() (auth.SchemeInfo, error) {
+	authnRequestData, err := s.generateAuthnRequest()
+	if err != nil {
+		return nil, err
+	}
+	r := request{}
+	//persist request in database
```
Comment (fsouza) [inline auth/saml/saml.go]: Please remove this comments as well :-)

File: auth/saml/saml_test.go [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
...TM2ZWVQMHpjaG8rMkN2eFZYVU1xcXU4ejVTL0s2YWh0aUJvYndjN1NzMkVKbkxzb1BmdzBaNHNVWGpVY1V2eEh2L1kvZU5BcWZvcU5BRVFRQUhjZzlaRTd4NUNRb2NUT1A0U0xmQlI5cDZRTnNFZ1BmbXpiN0RVVmRMZTk1MVZHeTlSV05OVzdqOS9yWWlaaWdoTDRWc0FpWlFuUUFhak9GSUNFQVo4MWR2Z3RRd0EzejBTUkZUMXpuYXJONnhrd081akVoN1lWRTI4RVJNNHUrSUs4R2Jmdjdoeis5ZUNqQVdVV0w0WHNZS0U2ZDkzemdXbjkxbXN4V0NqMTdIQkMwVVREcTFlSEtXRkRHc0hhbWoyRTRUVFBIN2MzSjJ3eHAyQnlhVlBFbHBXeXBlQ0tRTmxObXNtNWgrZWFxTnhRRjBBU2Y4cW9qU1Q0QStFbklDM3RQVFlkeGFBd0kxbUJtYUwxZ0tYb0R2dlEyelNGakc2QUpIY2RiWDNxZm5USzlHanJnTFBmWGxwRyt6V2VLck9LSmtpUEt0SEtZaEVaKzQyaUlIZU43MllTQ1hpM0NPVG5pN2tHZFFwWHJ5S240c0xzNWFkTjYza0pjam1mVXFqZW84aXh2SmlReTRuck5nd1kzbHJBZ085SVk0L3ZLMDhmNjNnUXZCdFpGaXJiM0RWUXdUUlZnT0ZzVG9TN2o2aVRvdUtsRG40UGtodFdqdVNYaWNKYm5IUDVUZEZ1ekRMSkNMMElodGM4OExOWXFFdnlLR01NUTJVamRvdHFXTmR5a2h5dXVweDhlWEIyUDh0YzZjTG5NVnZ2N2piT3hNZTRGVnZzVkFsemlreW9qb2FhK3lYTWVPWUZBV0M1YVVOUmJubEo4QWZWWExzYW40bXhUeEJRTmtLaTNJaFR4MWVNZ2RldURJczZzUjRqeHlQakxpYzlEYWJ4aHdRdEZEUmJQZmpuSURXUzlSM0Ztb1A3MWhjODFqQ3JyMVlnbjU3ZWhGa08zY1VpS3Ric0NLK2RLZjNBRFRxbXd5L1hIOW1LbjVJQUhRTTwveGVuYzpDaXBoZXJWYWx1ZT48L3hlbmM6Q2lwaGVyRGF0YT48L3hlbmM6RW5jcnlwdGVkRGF0YT48L3NhbWwyOkVuY3J5cHRlZEFzc2VydGlvbj48L3NhbWwycDpSZXNwb25zZT4="
```
Comment (fsouza) [inline auth/saml/saml_test.go]: Can you split this string in multiple lines? Something like:

```golang
b64Xml := "PD94..." +
        "continuation" +
        "continuation"
```

File: auth/saml/saml_test.go [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+func readRespFromFile() string {
+	buf := bytes.NewBuffer(nil)
+	f, _ := os.Open("testdata/resp.xml") // Error handling elided for brevity.
+	io.Copy(buf, f)                      // Error handling elided for brevity.
+	f.Close()
+	return string(buf.Bytes())
+}
+
+func (s *S) TestSamlCallbackDecodeXml(c *check.C) {
+	//Create request into database
```
Comment (fsouza) [inline auth/saml/saml_test.go]: Please remove this comment.

File: auth/saml/saml_test.go [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+	scheme := SAMLAuthScheme{}
+	//Parse xml
```
Comment (fsouza) [inline auth/saml/saml_test.go]: Please remove this comment as well.

File: auth/saml/saml_test.go (hunk 20 above)
Comment (fsouza) [inline auth/saml/saml_test.go]: Ah, BTW, please keep the comment above the string declaration, just so we can know what's this cryptographic string.

File: auth/saml/suite_test.go [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+
+func (s *S) SetUpSuite(c *check.C) {
+	s.server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		b, err := ioutil.ReadAll(r.Body)
+		c.Assert(err, check.IsNil)
+		s.bodies = append(s.bodies, string(b))
+		s.reqs = append(s.reqs, r)
+		w.Write([]byte(s.rsps[r.URL.Path]))
+	}))
+
```
Comment (fsouza) [inline auth/saml/suite_test.go]: Could you remove this blank line?

File: auth/saml/suite_test.go [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+	config.Set("auth:saml:idp-ssourl", "http://idp-service-url.com")
+	config.Set("auth:saml:idp-ssodescriptorurl", "Tsuru PaaS")
+	config.Set("auth:saml:idp-publiccert", "testdata/idp_pubcert.crt")
+	config.Set("auth:saml:sp-entityid", "tsuru.myservice.com")
+	config.Set("auth:saml:sp-sign-request", true)
+	config.Set("auth:saml:idp-sign-response", true)
+	config.Set("auth:saml:request-expire-seconds", 60)
+	config.Set("auth:user-registration", true)
+	config.Set("auth:saml:idp-attribute-user-identity", "eduPersonPrincipalName")
+
```
Comment (fsouza) [inline auth/saml/suite_test.go]: ditto

File: auth/saml/token.go [hunk 25]
```diff
@@ -0,0 +1,199 @@
...
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/tsuru/config"
+	"github.com/tsuru/tsuru/auth"
+	"github.com/tsuru/tsuru/db"
+	"github.com/tsuru/tsuru/permission"
+
```
Comment (fsouza) [inline auth/saml/token.go]: Please remove this blank line.

File: auth/saml/token.go [hunk 26]
```diff
@@ -0,0 +1,199 @@
...
+)
+
+const (
+	keySize           = 32
+	defaultExpiration = 7 * 24 * time.Hour
+)
+
+var (
+	tokenExpire time.Duration
+)
```
Comment (fsouza) [inline auth/saml/token.go]: You don't need the parenthesis for grouping here, as it's a single variable. Please do:

```go
var tokenExpire time.Duration
```

**Comment by fsouza** [comment]
@diego-araujo I think that the code looks good from a functional point of view, just some small changes in the code and it's good to merge.

Thank you very much for contributing!

**Comment by diego-araujo** [comment]
done!

**Comment by fsouza** [comment]
@diego-araujo thanks! There's one last comment here: https://github.com/tsuru/tsuru/pull/1324/files#diff-73d61840c61ef2ef3c5dc66747efaaabR11

Can you remove the import alias there? Then we're good to merge!

File: auth/saml/response.go (hunk 15 above)
Comment (diego-araujo) [inline auth/saml/response.go]: done :)

**Comment by fsouza** [comment]
Thank you!

**Pull request merged**